File: flutter_matter_ios/ios/Classes/Clusters/TemperatureCluster.swift
//
//  TemperatureCluster.swift
//  flutter_matter_ios
//
//  Created by Philipp Manstein on 24.10.23.
//

import Foundation
import OSLog
import Matter
import Flutter

class TemperatureCluster: FlutterMatterHostTemperatureClusterApi {
    
    
    private func getCluster(deviceId: Int64, endpointId: Int64) throws -> MTRBaseClusterTemperatureMeasurement {
        let controller = try MTRDeviceController.shared()
        let device = MTRBaseDevice(nodeID: NSNumber(value: deviceId), controller: controller)
        return MTRBaseClusterTemperatureMeasurement(device: device, endpointID: NSNumber(value: endpointId), queue: DispatchQueue.main)!
    }
    
    func readMeasuredValue(deviceId: Int64, endpointId: Int64, completion: @escaping (Result<Int64?, Error>) -> Void) {
        Task {
            do {
                let cluster = try getCluster(deviceId: deviceId, endpointId: endpointId)
                let result = try await cluster.readAttributeMeasuredValue();
                completion(Result.success(result.int64Value))
            }
            catch {
                os_log(.error, "Failed to read attribute MeasuredValue: \(error)")
                completion(Result.failure(FlutterError(code: "-1", message: "Failed to read attribute MeasuredValue", details: nil)))
            }
        }
    }
    
    func readMinMeasuredValue(deviceId: Int64, endpointId: Int64, completion: @escaping (Result<Int64?, Error>) -> Void) {
        Task {
            do {
                let cluster = try getCluster(deviceId: deviceId, endpointId: endpointId)
                let result = try await cluster.readAttributeMinMeasuredValue();
                completion(Result.success(result.int64Value))
            }
            catch {
                os_log(.error, "Failed to read attribute MinMeasuredValue: \(error)")
                completion(Result.failure(FlutterError(code: "-1", message: "Failed to read attribute MinMeasuredValue", details: nil)))
            }
        }
    }
    
    func readMaxMeasuredValue(deviceId: Int64, endpointId: Int64, completion: @escaping (Result<Int64?, Error>) -> Void) {
        Task {
            do {
                let cluster = try getCluster(deviceId: deviceId, endpointId: endpointId)
                let result = try await cluster.readAttributeMaxMeasuredValue();
                completion(Result.success(result.int64Value))
            }
            catch {
                os_log(.error, "Failed to read attribute MaxMeasuredValue: \(error)")
                completion(Result.failure(FlutterError(code: "-1", message: "Failed to read attribute MaxMeasuredValue", details: nil)))
            }
        }
    }
    
    func readTolerance(deviceId: Int64, endpointId: Int64, completion: @escaping (Result<Int64?, Error>) -> Void) {
        Task {
            do {
                let cluster = try getCluster(deviceId: deviceId, endpointId: endpointId)
                let result = try await cluster.readAttributeTolerance();
                completion(Result.success(result.int64Value))
            }
            catch {
                os_log(.error, "Failed to read attribute tolerance: \(error)")
                completion(Result.failure(FlutterError(code: "-1", message: "Failed to read attribute tolerance", details: nil)))
            }
        }
    }
}

File: flutter_matter_android/android/src/main/kotlin/net/grandcentrix/fluttermatter/clusters/TemperatureCluster.kt
package net.grandcentrix.fluttermatter.clusters

import chip.devicecontroller.ChipClusters
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import net.grandcentrix.fluttermatter.FlutterError
import net.grandcentrix.fluttermatter.FlutterMatterHostTemperatureClusterApi
import net.grandcentrix.fluttermatter.chip.ChipClient
import timber.log.Timber
import java.io.Closeable

class TemperatureCluster : FlutterMatterHostTemperatureClusterApi, Closeable {
    private lateinit var chipClient: ChipClient

    var activity: android.app.Activity? = null
        set(value) {
            if (value == activity) return
            field = value
            if (value != null) {
                chipClient = ChipClient(value!!)
            }
        }

    private suspend fun getCluster(
        deviceId: Long,
        endpointId: Long,
    ): ChipClusters.TemperatureMeasurementCluster {
        val devicePtr = chipClient.getConnectedDevicePointer(deviceId)
        return ChipClusters.TemperatureMeasurementCluster(devicePtr, endpointId.toInt())
    }

    private val scope =
        CoroutineScope(Dispatchers.IO + Job())

    override fun readMeasuredValue(
        deviceId: Long,
        endpointId: Long,
        callback: (Result<Long?>) -> Unit,
    ) {
        scope.launch {
            val cluster =
                try {
                    getCluster(deviceId, endpointId)
                } catch (e: Exception) {
                    callback(Result.failure(FlutterError("-1", "Can't connect to device!")))
                    null
                }

            cluster?.readMeasuredValueAttribute(
                object :
                    ChipClusters.TemperatureMeasurementCluster.MeasuredValueAttributeCallback {
                    override fun onSuccess(value: Int?) {
                        callback(Result.success(value?.toLong()))
                    }

                    override fun onError(ex: Exception) {
                        Timber.e(ex, "Failed to read attribute MeasuredValue")
                        callback(
                            Result.failure(
                                FlutterError(
                                    "-1",
                                    "Failed to read attribute MeasuredValue",
                                ),
                            ),
                        )
                    }
                },
            )
        }
    }

    override fun readMinMeasuredValue(
        deviceId: Long,
        endpointId: Long,
        callback: (Result<Long?>) -> Unit,
    ) {
        scope.launch {
            val cluster =
                try {
                    getCluster(deviceId, endpointId)
                } catch (e: Exception) {
                    callback(Result.failure(FlutterError("-1", "Can't connect to device!")))
                    null
                }

            cluster?.readMinMeasuredValueAttribute(
                object :
                    ChipClusters.TemperatureMeasurementCluster.MinMeasuredValueAttributeCallback {
                    override fun onSuccess(value: Int?) {
                        callback(Result.success(value?.toLong()))
                    }

                    override fun onError(ex: Exception) {
                        Timber.e(ex, "Failed to read attribute MinMeasuredValue")
                        callback(
                            Result.failure(
                                FlutterError(
                                    "-1",
                                    "Failed to read attribute MinMeasuredValue",
                                ),
                            ),
                        )
                    }
                },
            )
        }
    }

    override fun readMaxMeasuredValue(
        deviceId: Long,
        endpointId: Long,
        callback: (Result<Long?>) -> Unit,
    ) {
        scope.launch {
            val cluster =
                try {
                    getCluster(deviceId, endpointId)
                } catch (e: Exception) {
                    callback(Result.failure(FlutterError("-1", "Can't connect to device!")))
                    null
                }

            cluster?.readMaxMeasuredValueAttribute(
                object :
                    ChipClusters.TemperatureMeasurementCluster.MaxMeasuredValueAttributeCallback {
                    override fun onSuccess(value: Int?) {
                        callback(Result.success(value?.toLong()))
                    }

                    override fun onError(ex: Exception) {
                        Timber.e(ex, "Failed to read attribute MaxMeasuredValue")
                        callback(
                            Result.failure(
                                FlutterError(
                                    "-1",
                                    "Failed to read attribute MaxMeasuredValue",
                                ),
                            ),
                        )
                    }
                },
            )
        }
    }

    override fun readTolerance(
        deviceId: Long,
        endpointId: Long,
        callback: (Result<Long?>) -> Unit,
    ) {
        scope.launch {
            val cluster =
                try {
                    getCluster(deviceId, endpointId)
                } catch (e: Exception) {
                    callback(Result.failure(FlutterError("-1", "Can't connect to device!")))
                    null
                }

            cluster?.readToleranceAttribute(
                object :
                    ChipClusters.IntegerAttributeCallback {
                    override fun onSuccess(value: Int) {
                        callback(Result.success(value.toLong()))
                    }

                    override fun onError(ex: Exception) {
                        Timber.e(ex, "Failed to read attribute Tolerance")
                        callback(
                            Result.failure(
                                FlutterError(
                                    "-1",
                                    "Failed to read attribute Tolerance",
                                ),
                            ),
                        )
                    }
                },
            )
        }
    }

    override fun close() {
        scope.cancel()
    }
}
